### DISCUSSION TITLE
Question of understanding about encryption

### DISCUSSION THREAD

**Discussion by klixx23** [rustdesk/rustdesk#9172] [Q&A] [answered]
When I use the connection via the relay server, the message “Relayed and encrypted connection” is displayed in the top left-hand corner of the RustDesk window.

What exactly does this mean?
Is the data encrypted end-to-end between the clients or is it only encrypted end-to-end between the client and the relay server?

### Example:
**Scenario 1 (Client 2 Client):**
Client A (encrypted with PublicKey Client B)-> Relay server forwards the encrypted data directly to Client B -> Client B decrypts the data with its private key

**Scenario 2 (Client 2 Relay-Server 2 Client):**
Client A encrypts the data with the public key of the relay server (key in the RustDesk client settings) -> Relay server decrypts the data from client A and then encrypts it with the public key of client B -> Client B then decrypts the data with its private key.

Which scenario applies if the data is sent via the relay server?

Thank you

**Comment by huss2342** [comment, marked as answer]
After digging through the code, especially in the 'rendezvous_mediator.rs' and 'client.rs' files, I think that Scenario 1 is what's happening when data is sent through the relay server in RustDesk.

When you see the "Relayed and encrypted connection" message, it means your connection is being routed through a relay server because a direct connection couldn’t be established. But your data is still encrypted end-to-end between the clients.

From what I found in the code, RustDesk makes it so that even when a relay server is involved, your data stays encrypted from one client to the other. The relay server’s job is just to pass along these encrypted data packets without being able to decrypt or read them.

In 'rendezvous_mediator.rs', when setting up a relay connection, there's a 'secure' parameter that’s set to true, meaning that the connection is intended to be encrypted. Meanwhile, the 'client.rs' file shows how secure connections are established directly between the clients using their public keys. This creates a symmetric encryption key that only the two clients know, which is used to encrypt the data.

I didn’t find any evidence that the relay server is involved in the encryption or decryption process—it just forwards the already-encrypted data.

So, to directly answer your question based on what I figured out from the code: When you see "Relayed and encrypted connection," it means your connection is being relayed through a server, but the data is still securely encrypted end-to-end between the clients (Scenario 1). The relay server can’t read or tamper with your communication, it just helps establish the connection when a direct link isn’t possible.

**Reply by klixx23** [reply]
Thank you very much for the detailed feedback. 👍
